dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -82,9 +84,100 @@ public Pattern getParameterPattern() {
+        try (ResultSet rs = checkForFunction ? dbmd.getFunctions(null, qn.getSchemaName(), qn.getObjectName()) :
+                                               dbmd.getProcedures(null, qn.getSchemaName(), qn.getObjectName())) {
+            return rs.next();
+        }
+    }
+
+    private Direction getParamDirection(int columnType, boolean isFunction) {
+        if (isFunction) {
+            switch (columnType) {
+                case DatabaseMetaData.functionColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.functionReturn:
+                    return Direction.RETURN_VALUE;
+            }
+        } else {
+            switch (columnType) {
+                case DatabaseMetaData.procedureColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.procedureColumnInOut:
+                    return Direction.INPUT_OUTPUT;
+                case DatabaseMetaData.procedureColumnOut:
+                    return Direction.OUTPUT;
+            }
+        }
+        throw new UnsupportedOperationException("Type " + columnType
+                      + " is not a supported type for "
+                      + (isFunction ? "function" : "procedure") + " parameters");
+    }
+
+    private DbParameterAccessor crtProcDbParamAccessor(ResultSet rs,
+                                                       boolean isFunction)
+                                                       throws SQLException {
+        String paramName = defaultIfNull(rs.getString("COLUMN_NAME"), "");
+        Direction paramDirection = getParamDirection(rs.getInt("COLUMN_TYPE"), isFunction);
+        String paramTypeName = rs.getString("TYPE_NAME");
+        int position = rs.getInt("ORDINAL_POSITION"); // We rely upon Derby returning 0 for function return value.
+        return createDbParameterAccessor(paramName, paramDirection,
+                   typeMapper.getJDBCSQLTypeForDBType(paramTypeName), getJavaClass(paramTypeName),
+                   (position > 0 ? position : -1));
+    }
+
+    private ResultSet getObjectParameters(DbObjectName qn, boolean isFunction)
+                                                            throws SQLException {
+        DatabaseMetaData dbmd = getConnection().getMetaData();
+        return isFunction ? dbmd.getFunctionColumns(null, qn.getSchemaName(), qn.getObjectName(), "%")
+                          : dbmd.getProcedureColumns(null, qn.getSchemaName(), qn.getObjectName(), "%");
+    }
+
+    private Map<String, DbParameterAccessor> getParamMap(DbObjectName qn,
+                                                         boolean isFunction)
+                                                         throws SQLException {
+        Map<String, DbParameterAccessor> allParams = new HashMap<>();
+        try (ResultSet rs = getObjectParameters(qn, isFunction)) {
+            while (rs.next()) {
+                String paramName = defaultIfNull(rs.getString("COLUMN_NAME"), "");
```

Comment:
Can we avoid that code duplication (the same line is in `crtProcDbParamAccessor`)?